ibtk/src/utilities/SnapshotCache.cpp

```diff
@@ -0,0 +1,247 @@
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "ibtk/SnapshotCache.h"
+#include "ibtk/ibtk_utilities.h"
+
+#include "CartesianGridGeometry.h"
+#include "RefineAlgorithm.h"
+#include "RefineOperator.h"
+#include "VariableDatabase.h"
+
+#include <algorithm>
+
+#include "ibtk/namespaces.h" // IWYU pragma: keep
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBTK
+{
+template <class VariableType>
+SnapshotCache<VariableType>::SnapshotCache(std::string object_name, Pointer<Database> input_db)
+    : d_object_name(std::move(object_name))
+{
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    d_ctx = var_db->getContext(d_object_name + "::context");
+
+    if (input_db)
+    {
+        d_snapshot_refine_type = input_db->getStringWithDefault("refine_type", d_snapshot_refine_type);
+        if (input_db->keyExists("gcw")) input_db->getIntegerArray("gcw", &d_gcw[0], NDIM);
+        d_depth = input_db->getIntegerWithDefault("depth", d_depth);
+    }
+}
+
+template <class VariableType>
+SnapshotCache<VariableType>::~SnapshotCache()
+{
+    clearSnapshots();
+}
+
+template <class VariableType>
+void
+SnapshotCache<VariableType>::clearSnapshots()
+{
+    // Clear the map.
+    d_idx_time_map.clear();
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    // Deallocate all patch data.
+    for (const auto& idx : d_snapshot_idxs)
+    {
+        const Pointer<PatchHierarchy<NDIM> >& hierarchy = d_snapshot_hierarchies[idx];
+        int coarsest_ln = 0;
+        int finest_ln = hierarchy->getFinestLevelNumber();
+        for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+        {
+            Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(ln);
+            if (level->checkAllocated(idx)) level->deallocatePatchData(idx);
+        }
+        // Delete variable from variable database
+        var_db->removePatchDataIndex(idx);
+    }
+
+    // Clear snapshots
+    d_snapshot_hierarchies.clear();
+    d_snapshot_idxs.clear();
+    d_snapshot_vars.clear();
+    d_num_snapshots_stored = 0;
+}
+
+template <class VariableType>
+void
+SnapshotCache<VariableType>::setSnapshot(const int u_idx,
+                                         const double time,
+                                         Pointer<PatchHierarchy<NDIM> > current_hierarchy)
+{
+    ++d_num_snapshots_stored;
+    // Push back a new variable
+    Pointer<VariableType> snapshot_var =
+        new VariableType(d_object_name + "::SnapshotVar_" + std::to_string(d_num_snapshots_stored), d_depth);
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    int snapshot_idx = var_db->registerVariableAndContext(snapshot_var, d_ctx, d_gcw);
+    d_snapshot_idxs.push_back(snapshot_idx);
+    d_snapshot_vars[snapshot_idx] = snapshot_var;
```

Comment:
We can use the same variable from the provided index, but we do need a new context. We need to specify a potentially different ghost cell width.

Reply:
Got it. If we have a separate hierarchy per snapshot then would just reusing the same patch data index and variable on  each hierarchy work?

Reply:
Because we have our own copy of the patch hierarchy, I suppose we could avoid the `VariableDatabase` altogether and manage the `PatchDescriptor` and `PatchDataFactory` objects directly.

Reply:
Avoiding `VariableDatabase` sounds like a good thing to do, particularly for "internal variables".

Reply:
My understanding is that `PatchData` is allocated by a `PatchDataFactory` object, which is in turn handled by the `PatchDescriptor` object. This object is passed from the patch hierarchy to the patch level to the individual patches so that each patch has the same `PatchDescriptor` object. On construction, the `PatchHierarchy` object uses the `PatchDescriptor` created by the `VariableDatabase`, and there's no method to change which `PatchDescriptor` a `PatchHierarchy` uses.

I see two paths forward:

1. Avoid `VariableDatabase` and `PatchHierarchy` and manage the snapshot `PatchLevel`'s directly. This class would have to carry around the different `PatchLevel` objects and a `PatchDescriptor` object, but we wouldn't have to use `VariableDatabase`. Writing restart files might also be a bit tricky.
2. Use `VariableDatabase` to create one variable and context pairs per instance of `SnapshotCache`.

@boyceg and @drwells, do you have an opinion?

Reply:
Option 2 seems like the simplest.

However, from the standpoint of having an accurate representation of the data, would it make sense to actually store the `PatchHierarchy` for the snapshotted-configuration, so that the snapshot is not slowly degraded by dynamic mesh refinement/coarsening? To do this, it would be useful to have an internal `PatchHierarchy`/`PatchLevel` that _only_ has data allocated for the captured data.

Reply:
That's what this is currently set up to do. We store a snapshot of the patch hierarchy and only allocate data for the captured patch index of that snapshot.

We were trying to avoid having to use the `VariableDatabase` to get a patch index to allocate data on the snapshotted patch hierarchy. The issue is an instance of `PatchHierarchy` can only use the `PatchDescriptor` object managed by `VariableDatabase`. So the options are we store a `PatchHierarchy` object (which contains a snapshot of the patch configuration) and use `VariableDatabase` to get a patch index to store data on. Or we store the `PatchLevel`s ourselves and manage our own `PatchDescriptor` object to give us a patch index.

Reply:
I think the best way forward is to just create one `PatchHierarchy` per snapshot. I think we all agree on this.

In general its just not possible to avoid the variable database - at some point we need to map between indices and `Variable` objects and that is what it does.

Reply:
If we want to use a `PatchHierarchy` object, then we have to use the `VariableDatabase`. We could add our own index to the `PatchDescriptor` object, but that's functionally the same as using `VariableDatabase`.

Reply:
OK --- I think it seems easiest to manage `PatchHierarchy` and use `VariableDatabase`.